andela/Titan-Devs
# #163518705 Role based access control

#### What does this PR do?
- Add roles-based access control on routes
#### Description of Task to be completed?
- Protect routes with  `create, read, delete, update` Permissions
#### How should this be manually tested?
- After cloning the repository, check out `ft-role-based-access-control-#23163518705` branch
- Make sure to read `README.md` file to get the setup instructions
- Using any `API client tool(usually POSTMAN)`,
- Create the user and login to get the `access-token`
- Make sure to pass the token under `request headers`. It should be {`Authorization`: `Bearer <access-token>`} KEY, VALUE PAIR,
- Type this command `yarn db:migrate:seed ` to get super admin
Apply the following process:
#### Perform CRUD operation on Permission object using the follwing endpoints
- Create  a role `POST /api/v1/roles` by providing  `name` and `description`
- Read all roles `GET /api/v1/roles`
- Read a role `GET /api/v1/roles/:roleId` by providing `roleId` as url parameter 
- Update a role `PUT /api/v1/role/:roleId` by providing `roleId` as url parameter 
- Delete a role `DELETE /api/v1/role/:roleId`by providing `roleId` as url parameter 
#### Update a User object
- Assign role to a user using `PUT /api/v1/users/:username/roles/roleId`, by providing `username` and `role Id`.
#### Perform CRUD operation on Role object using the follwing endpoints
- CREATE/GRANT permission to a role `POST /api/v1/permissions/:roleId` provide `roleId` as parameter and `{resource, createPermission, readPermission, updatePermission, deletePermission`} int the body except `resource` other are ***OPTIONAL***
- Read all permissions `GET /api/v1/permissions`
- Update a permission `PUT /api/v1/permissions/:permissionId`
- Delete a permission `DELETE /api/v1/permissions/:permissionId`

#### Any background context you want to provide?
- The permission constrol is being held inside checkAuth middleware
#### What are the relevant pivotal tracker stories?
- #163518705
